aten/src/ATen/native/cpu/IndexKernel.cpp

```diff
@@ -525,6 +528,155 @@ void cpu_hflip_vec(at::TensorIterator& iter) {
   iter.cast_outputs();
 }
 
+void cpu_hflip_channels_last_uint8_ch3(at::TensorIterator& iter) {
+
+  auto loop2d = [&](char** base, const int64_t *strides, int64_t size0, int64_t size1) {
+
+    // Here ntensors is defined for output and 1 input. But tensor iterator has defined output, input
+    // and restrided_input (see aten/src/ATen/native/TensorTransformations.cpp#L64-L66) but we use only
+    // output and input.
+    static constexpr int ntensors = 2;
+    const int64_t *outer_strides = &strides[3];
+    const int64_t stride = strides[0];
+
+    TORCH_INTERNAL_ASSERT(strides[0] == strides[1]);
+    TORCH_INTERNAL_ASSERT(stride == 1);
+    TORCH_INTERNAL_ASSERT(outer_strides[0] == -3);
+    TORCH_INTERNAL_ASSERT(outer_strides[1] == 3);
+
+    char* C10_RESTRICT data[ntensors] = {base[0], base[1]};
+    const int64_t size = size0 * size1;
+
+    int64_t i = 0;
+
+#ifdef CPU_CAPABILITY_AVX2
+
+    if (size0 == 3) {
+      constexpr auto vec_size = 256 / (8 * sizeof(uint8_t));
+      const auto proc_vec_size = 2 * ((vec_size / 2) / size0);
+      const auto delta = vec_size - proc_vec_size * size0;
+      __m256i data_vec, reversed_vec;
+
+      // Data: (1 2 3) (4 5 6) (7 8 9) (10 11 12) (13 14 15) (16 17 18) (19 20 21) (22 23 24) (25 26 27) (28 29 30) (31 32 33)
+      // load by 2 parts
+      // R = [ (1 2 3) (4 5 6) (7 8 9) (10 11 12) (13 14 15) (16 | (16 17 18) (19 20 21) (22 23 24) (25 26 27) (28 29 30) (31 ]
+      // flip(R) ->
+      // R = [ 31 (28 29 30) (25 26 27) (22 23 24) (19 20 21) (16 17 18) | 16 (13 14 15) (10 11 12) (7 8 9) (4 5 6) (1 2 3) ]
+      // Write in 2 parts
+
+      // Output pointer:                                                                                                       v
+      //                (X X X)  (X X X)    (X X X)    (X X X)    (X X X)    (X X X)    (X X X)    (X X X)    (X X X) (X X X) (X X X)
+
+      // Output part 1: (X X X)  (X X X)    (X X X)    (X X X)    (X X X)    (X X 16)   (13 14 15) (10 11 12) (7 8 9) (4 5 6) (1 2 3)
+      // Output part 2: (X X 31) (28 29 30) (25 26 27) (22 23 24) (19 20 21) (16 17 18) (13 14 15) (10 11 12) (7 8 9) (4 5 6) (1 2 3)
+
+      // Shuffle mask for 3 channels
+      __m256i mask =_mm256_set_epi8(
+          2, 1, 0, 5, 4, 3, 8, 7, 6, 11, 10, 9, 14, 13, 12, 15, // first 128-bit lane
+          2, 1, 0, 5, 4, 3, 8, 7, 6, 11, 10, 9, 14, 13, 12, 15  // second 128-bit lane
+      );
+
+      // shift output pointer by size0
+      if (size > proc_vec_size * size0) {
+        data[0] += size0 * stride;
+      }
+
+      for (; i < size - proc_vec_size * size0; i += proc_vec_size * size0) {
```

Comment:
This can read off the end, right? It reads up to `i + vec_size / 2 + proc_vec_size / 2`